src/NServiceBus.Core/Performance/Metrics/ReceiveDiagnosticsBehavior.cs

```diff
@@ -0,0 +1,46 @@
+namespace NServiceBus.Performance.Metrics
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Metrics;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class ReceiveDiagnosticsBehavior : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
+    {
+        static readonly Meter NServiceBusMeter = new Meter(
+            NServiceBusDiagnosticsInfo.InstrumentationName,
+            NServiceBusDiagnosticsInfo.InstrumentationVersion);
+
+        static readonly Counter<long> TotalProcessedSuccessfully =
+            NServiceBusMeter.CreateCounter<long>("messaging.successes", "Total number of messages processed successfully by the endpoint.");
+
+        static readonly Counter<long> TotalFetched =
+            NServiceBusMeter.CreateCounter<long>("messaging.fetches", "Total number of messages fetched from the queue by the endpoint.");
+
+        static readonly Counter<long> TotalFailures =
+            NServiceBusMeter.CreateCounter<long>("messaging.failures", "Total number of messages processed unsuccessfully by the endpoint.");
+
+        public async Task Invoke(IIncomingPhysicalMessageContext context, Func<IIncomingPhysicalMessageContext, Task> next)
+        {
+            context.MessageHeaders.TryGetMessageType(out var _);
+
+            //TODO: do we need to tag the message type, or any other data? 
+            var tags = Array.Empty<KeyValuePair<string, object>>();
+
+            TotalFetched.Add(1, tags);
+
+            try
+            {
+                await next(context).ConfigureAwait(false);
+            }
+            catch (Exception ex) when (!ex.IsCausedBy(context.CancellationToken))
```

Comment:
This means that cancellations are not counted as successes or failures but they are counted as fetches. Given that we're rethrowing anyway, should a cancellation be counted as a failure here?

Reply:
I'm not sure. This excludes messages aborted by cancellation tokens from being counted as success messages (which sounds correct to me). The exclusion is something we do all over the place (e.g. UnitOfWorkBehavior).